Better align 1q euler decomposition output to multiples of π/2

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit updates the euler decomposer to better align the parameter outputs to be multiples of π/2 when possible to enable output circuits to be run with a Clifford simulator.

Previously the angle was calculated as:

```
phi = angle(u11 * (det**(-0.5))) + angle(u10 * (det**(-0.5)))
```

which has been replaced with:

```
phi = angle(u11) + angle(u10) - angle(det)
```

The algebra to get the new expression is:

```
phi = angle(u11 * (det**(-0.5))) + angle(u10 * (det**(-0.5)))
    # assumes particular phase convention so not guaranteed to be the
    # valid interpretation (though it probably is).
    = angle(u11) - 0.5*angle(det) + angle(u10) - 0.5*angle(det)
    = angle(u11) + angle(u10) - angle(det)
```

A couple of tests are updated as they were testing for exact circuit outputs and the synthesis is returning equivalent by different results in those cases.

### Details and comments

Co-authored-by: aeddins-ibm <60495383+aeddins-ibm@users.noreply.github.com>

mtreinish committed:
Better align 1q euler decomposition output to multiples of π/2

This commit updates the euler decomposer to better align the parameter
outputs to be multiples of π/2 when possible to enable the circuit to
be run under a clifford simulator.

Previously the angle was calculated as:

```
phi = angle(u11 * (det**(-0.5))) + angle(u10 * (det**(-0.5)))
```

which has been replaced with:

```
phi = angle(u11) + angle(u10) - angle(det)
```

The algebra to get the new expression is:

```
phi = angle(u11 * (det**(-0.5))) + angle(u10 * (det**(-0.5)))
    # assumes particular phase convention so not guaranteed to be the
    # valid interpretation (though it probably is).
    = angle(u11) - 0.5*angle(det) + angle(u10) - 0.5*angle(det)
    = angle(u11) + angle(u10) - angle(det)
```

A couple of tests are updated as they were testing for exact circuit
outputs and the synthesis is returning equivalent by different results
in those cases.

Co-authored-by: aeddins-ibm <60495383+aeddins-ibm@users.noreply.github.com>

mtreinish committed:
Merge branch 'main' into better-align-1q-synth-towards-clifford

jakelishman commented:
`f64::atan2` returns an angle in `(-pi, pi]`, and `Complex64::sqrt` calculates the principal root, so satisfies `-pi / 2 <= x.arg() < pi / 2`, so I think the phase convention mentioned in the algebraic comment _is_ satisfied.

jakelishman commented on crates/accelerate/src/euler_one_qubit_decomposer.rs:
```diff
@@ -688,15 +688,16 @@ fn mod_2pi(angle: f64, atol: f64) -> f64 {
-    let phi = phiplambda2 + phimlambda2;
-    let lam = phiplambda2 - phimlambda2;
+    let ang1 = complex_phase(mat[[1, 1]]);
+    let ang2 = complex_phase(mat[[1, 0]]);
```
fwiw, the `complex_phase` function we've defined here is just `Complex64::arg`.

mtreinish replied:
Hah, I must have missed that in the `num-complex` docs at the time. I'll push up a follow up PR that removes it in favor of `Complex64::arg`

mtreinish replied:
Actually I assumed it was used in more places. but it's only used in this function. I'll adjust it as part of this PR.

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/10789/commits/72af34f87b0bdd2a44bd76ff554e9ffc1ed68330

jakelishman commented on crates/accelerate/src/euler_one_qubit_decomposer.rs:
```diff
@@ -688,15 +688,16 @@ fn mod_2pi(angle: f64, atol: f64) -> f64 {
     let phase = -complex_phase(coeff);
     let tmp_1_0 = (coeff * mat[[1, 0]]).abs();
     let tmp_0_0 = (coeff * mat[[0, 0]]).abs();
     let theta = 2. * tmp_1_0.atan2(tmp_0_0);
```
I didn't notice this at the time we merged this, but since `tmp_1_0` and `tmp_0_0` are both guaranteed positive real and only used in an `atan2` call, multiplying by `coeff` has no effect (it immediately cancels), and I think this could just be `theta = 2. * mat[[1, 0]].abs().atan2(mat[[0, 0]].abs())`.

jakelishman replied:
Come to think of it, if we do that, I don't even think we need to calculate the reciprocal square root, which is actually a quite nice saving - the `phase` being calculated as `-arg(det ** -0.5)` reduces to `0.5 * arg(det)` under the same assumptions about the phase convention.

jakelishman replied:
I suppose technically it has an effect if `coeff` is 0 or infinite, but that would require an infinite or zero determinant respectively, which can't happen for unitary matrices.

mtreinish replied:
I made these simplification in: https://github.com/Qiskit/qiskit/pull/10789/commits/4c08c0adf46d060c14b581561ad1a51c59199e58

aeddins-ibm replied:
since `mat[[1,0]].abs()` and `mat[[0,0]].abs()` are both real, could it make sense to call `atan` instead of `atan2`? Just in case that's any faster

jakelishman replied:
`mat[[0, 0]]` could still be zero I think?  We wouldn't be able to produce the argument to pass to `atan` by division if so, in which case we'd just need to do the branching that `atan2` will do anyway.

aeddins-ibm replied:
oh right, that makes sense.

jakelishman reviewed (commented):
In principle this looks fine to me.  The QPY backwards compatibility tests could be a nuisance, potentially.

mtreinish committed:
Simplify calculation

Given the same phase convention the calculation of theta and phase can
be simplified to remove the need of computing the coeff. This commit
implements this change which should speed up the calculation slighty as
we're doing less work to reach an equivalent result.

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

mtreinish committed:
Remove complex_phase function

The complex phase function previously defined in
euler_one_qubit_decomposer.rs was duplicated with an existing `arg()`
method of the Complex<T> type from num-complex. This commit removes the
duplicated definition in favor of the built-in method.

mtreinish committed:
Try removing rz gate from qpy tests to side step differing decomposition of controlled gate

jakelishman reviewed (dismissed):
Assuming the QPY backwards compatibility tests pass now, I'm happy with this.  I wouldn't mind having @aeddins-ibm or @levbishop check me on the extra mathematical bits I suggested to simplify the decomposition calculation, though.

mtreinish commented:
The tests are currently blocked because of running out of disk space, this should be fixed by #10794

aeddins-ibm commented on crates/accelerate/src/euler_one_qubit_decomposer.rs:
```diff
@@ -688,15 +683,13 @@ fn mod_2pi(angle: f64, atol: f64) -> f64 {
+    let ang1 = mat[[1, 1]].arg();
+    let ang2 = mat[[1, 0]].arg();
+    let phi = ang1 + ang2 - det.arg();
+    let lam = ang1 - ang2;
```
Can remove one call to arg() since `det` itself is not needed?
```suggestion
    let det_arg = det_one_qubit(mat).arg();
    let phase = 0.5 * det_arg;
    let theta = 2. * mat[[1, 0]].abs().atan2(mat[[0, 0]].abs());
    let ang1 = mat[[1, 1]].arg();
    let ang2 = mat[[1, 0]].arg();
    let phi = ang1 + ang2 - det_arg;
    let lam = ang1 - ang2;
```

jakelishman replied:
In theory LLVM should detect that `Complex64::arg` is a pure function and optimise out all but the first call, but it doesn't hurt to be more explicit.

aeddins-ibm replied:
oh, that's cool, thanks for educating me. Probably not worth the edit then.

jakelishman replied:
Oh, I still think it's a good idea to put in - no particular reason to rely on LLVM noticing it and doing it when it's easy for us to just be explicit.

mtreinish committed:
Update crates/accelerate/src/euler_one_qubit_decomposer.rs

Co-authored-by: aeddins-ibm <60495383+aeddins-ibm@users.noreply.github.com>

mtreinish committed:
Merge branch 'main' into better-align-1q-synth-towards-clifford

jakelishman commented:
(just waiting for @levbishop to either check the maths or say that he can't)

levbishop commented:
Thanks for the ping I'll take a look

jakelishman commented:
I think Lev's a bit busy to look, but since Andrew did, I'll just move this to merge.
